aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -0,0 +1,514 @@
+    serial_for_each(loop, base_ptrs, inner_strides, begin, end - begin);
+  });
+}
+
+void TensorIterator::serial_for_each(loop_t loop, ArrayRef<char*> base_ptrs, IntList inner_strides, int64_t start, int64_t n) {
+  if (ndim() <= 1) {
+    auto ptrs = get_data_ptrs(base_ptrs, { start });
+    loop(ntensors(), ptrs.data(), inner_strides.data(), n);
+  } else {
+    auto counter = make_counter(start);
+    while (n > 0) {
+      auto ptrs = get_data_ptrs(base_ptrs, counter);
+      int64_t loop_size = std::min(n, shape_[0] - counter[0]);
+      loop(ntensors(), ptrs.data(), inner_strides.data(), loop_size);
+      n -= loop_size;
+      if (n == 0) break;
+      increment_counter(counter, loop_size);
+    }
+  }
+}
+
+bool TensorIterator::is_trivial_1d() const {
+  // TODO: check for casting once it's supported
+  return ndim() == 1;
+}
+
+bool TensorIterator::is_scalar(int arg) const {
+  const auto& stride = operands_[arg].stride_;
+  for (int i = 0; i < ndim(); i++) {
+    if (stride[i] != 0 && shape_[i] != 1) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool TensorIterator::is_cpu_scalar(int arg) const {
+  return is_scalar(arg) && operands_[arg].tensor_->type().backend() == at::kCPU;
+}
+
+void* TensorIterator::data_ptr(int arg) const {
+  return operands_[arg].data_;
+}
+
+void TensorIterator::remove_operand(int arg) {
+  operands_.erase(operands_.begin() + arg);
+}
+
+void TensorIterator::narrow(int dim, int64_t start, int64_t size) {
+  AT_ASSERT(dim < ndim() && size >= 1);
+  shape_[dim] = size;
+  for (auto& op : operands_) {
+    op.data_ = ((char*)op.data_) + op.stride_[dim] * start;
+  }
+  if (size == 1) {
+    coalesce_dimensions();
+  }
+}
+
+std::unique_ptr<TensorIterator> TensorIterator::binary_op(const Tensor& a, const Tensor& b, Tensor& out) {
+  auto builder = TensorIterator::Builder();
+  builder.add_output(out);
+  builder.add_input(a);
+  builder.add_input(b);
+  return builder.build();
+}
+
+TensorIterator TensorIterator::reduce_op(const Tensor& a, IntList dims) {
+  return TensorIterator();
+}
+
+void TensorIterator::mark_outputs() {
+  for (int i = 0; i < num_outputs_; i++) {
+    operands_[i].is_output_ = true;
+    auto output = *operands_[i].tensor_;
+    if (!output.defined()) continue;
+
+    // check if output is also an input
+    for (int arg = num_outputs_; arg < ntensors(); arg++) {
+      auto input = *operands_[arg].tensor_;
+      if (output.get() == input.get()) {
+        operands_[i].is_read_write_ = true;
+      }
+    }
+  }
+}
+
+void TensorIterator::compute_shape() {
+  for (auto& op : operands_) {
+    if (!op.tensor_->defined()) continue;
+
+    // For now, don't include output tensors that are not also input tensors.
+    // This preserves the legacy behavior where torch.add(..., out=dst) resizes
+    // the destination tensor.
+    if (op.is_output_ && !op.is_read_write_) continue;
+
+    auto shape = op.tensor_->sizes();
+    if (shape_.empty()) {
+      shape_ = shape;
```

Comment:
why do we need to do this?  doesn't infer_size handle this?  Or maybe it will after https://github.com/pytorch/pytorch/pull/9209?